File: internal/api/sse.go
package api

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/eyupio/zoomies/internal/auth"
	"github.com/eyupio/zoomies/internal/backend"
	"github.com/eyupio/zoomies/internal/controller"
	"github.com/eyupio/zoomies/internal/events"
	"github.com/eyupio/zoomies/internal/store"
)

// heartbeatInterval is how often a live stream sends something.
//
// Twenty seconds is comfortably inside the idle timeout of every proxy and load
// balancer anyone puts in front of this, which is what keeps a quiet fleet's
// event stream from being closed under the operator every minute. The UI's
// watchdog is set from the same number.
const heartbeatInterval = 20 * time.Second

// sseWriter is one Server-Sent Events response.
//
// It exists so that the two streams -- fleet events and a runner's log tail --
// cannot disagree about framing, flushing or heartbeats, which is the usual way
// one of two SSE endpoints ends up subtly buffered.
type sseWriter struct {
	w  http.ResponseWriter
	rc *http.ResponseController
}

// startSSE writes the headers and returns the stream, or nil when the client
// cannot be streamed to.
func startSSE(w http.ResponseWriter, r *http.Request) *sseWriter {
	h := w.Header()
	h.Set("Content-Type", "text/event-stream; charset=utf-8")
	h.Set("Cache-Control", "no-cache, no-transform")
	h.Set("Connection", "keep-alive")
	// nginx buffers proxied responses by default, which turns a live stream
	// into one that arrives in lumps minutes late. This is the header that
	// turns that off, and it costs nothing when there is no nginx.
	h.Set("X-Accel-Buffering", "no")

	s := &sseWriter{w: w, rc: http.NewResponseController(w)}
	// A stream has no length and no deadline. The server's WriteTimeout is
	// already 0 for exactly this reason; clearing it here as well means a
	// caller that serves this handler from a listener of their own cannot cut
	// the stream off by setting one.
	_ = s.rc.SetWriteDeadline(time.Time{})

	w.WriteHeader(http.StatusOK)
	// The first flush is what makes the browser's EventSource fire onopen,
	// which the UI's connection indicator is driven by.
	s.flush()
	return s
}

// event writes one named event with its id and JSON payload. An empty id
// leaves the client's last id where it was, which is what a heartbeat or a
// log chunk wants.
func (s *sseWriter) event(kind string, id string, data []byte) error {
	var b strings.Builder
	if id != "" {
		b.WriteString("id: ")
		b.WriteString(id)
		b.WriteByte('\n')
	}
	if kind != "" {
		b.WriteString("event: ")
		b.WriteString(kind)
		b.WriteByte('\n')
	}
	if len(data) == 0 {
		data = []byte("{}")
	}
	// A payload containing a newline has to be split across data: lines or the
	// frame ends early and the client sees truncated JSON.
	for line := range strings.SplitSeq(string(data), "\n") {
		b.WriteString("data: ")
		b.WriteString(strings.TrimSuffix(line, "\r"))
		b.WriteByte('\n')
	}
	b.WriteByte('\n')
	if _, err := io.WriteString(s.w, b.String()); err != nil {
		return err
	}
	s.flush()
	return nil
}

// comment writes a heartbeat. A comment frame keeps the connection alive
// without the client having to know about a keep-alive event kind.
func (s *sseWriter) comment(text string) error {
	if _, err := fmt.Fprintf(s.w, ": %s\n\n", text); err != nil {
		return err
	}
	s.flush()
	return nil
}

// retry tells the browser how long to wait before reconnecting on its own.
func (s *sseWriter) retry(d time.Duration) error {
	if _, err := fmt.Fprintf(s.w, "retry: %d\n\n", d.Milliseconds()); err != nil {
		return err
	}
	s.flush()
	return nil
}

func (s *sseWriter) flush() { _ = s.rc.Flush() }

// ---------------------------------------------------------------------------
// GET /api/v1/events
// ---------------------------------------------------------------------------

// endIfRevoked re-resolves the credential this stream was opened with and, when
// it no longer stands, ends the stream with a frame saying so.
//
// A live stream is authorised once, at the moment it is opened, and then held
// for as long as the tab is. Without this, signing out, revoking the token,
// disabling the account or letting the session expire leaves the stream running
// -- the fleet's every movement still arriving at a browser whose credential
// the operator has just taken away, until something else happens to break the
// connection. Every other route re-checks on every request; a stream's
// heartbeat is the closest thing it has to one, so it re-checks there.
//
// The reason is deliberately the same whichever way the credential failed: an
// ended stream tells a client to stop and sign in again, and which of the four
// happened is not something the far end can do anything different about.
func (s *Server) endIfRevoked(stream *sseWriter, r *http.Request, action auth.Action) bool {
	id, err := s.resolveIdentity(r)
	if err == nil && auth.Allowed(id, action) {
		return false
	}
	_ = stream.event("end", "", []byte(`{"reason":"this stream's credential is no longer valid; sign in again"}`))
	return true
}

// handleEvents streams every change the operator would want to see.
//
// There is no polling anywhere in the UI, so this endpoint is the whole of the
// dashboard's liveness. It replays from Last-Event-ID on reconnect, which is
// what makes a dropped connection invisible rather than a hole in the history.
func (s *Server) handleEvents(w http.ResponseWriter, r *http.Request) {
	bus := s.ctrl.Events()
	replay, sameEpoch := lastEventID(r, bus)
	opts := events.SubscribeOptions{
		Kinds:       parseKinds(r.URL.Query().Get("kinds")),
		TopicPrefix: strings.TrimSpace(r.URL.Query().Get("topic")),
		Replay:      replay,
		Incomplete:  replay > 0 && !sameEpoch,
	}

	ctx := r.Context()
	// The subscription is unsubscribed by cancelling the context it was made
	// with, which the bus watches and acts on. That is one owner of the
	// teardown rather than two: closing it here as well would mean this
	// handler and the bus's own watcher racing to do the same thing, and the
	// point of the cancel is that it happens on every exit from this function,
	// including a write failure to a browser that has already gone.
	subCtx, unsubscribe := context.WithCancel(ctx)
	defer unsubscribe()
	sub := bus.Subscribe(subCtx, opts)

	stream := startSSE(w, r)
	_ = stream.retry(2 * time.Second)
	// Decided once, for this connection. A pool frame carries the same shape
	// the GET does, and the GET withholds env values from a caller who could
	// not have set them; a stream that handed the same operator's registry
	// password to a viewer would just be the longer way round to the same
	// disclosure.
	redactPoolEnv := !auth.Allowed(Identity(r.Context()), auth.ActionPoolsWrite)

	// An immediate heartbeat lets the client start its stall watchdog without
	// waiting twenty seconds to learn the stream works.
	if err := stream.comment("connected"); err != nil {
		return
	}
	if !sub.Complete {
		// The client asked to resume from an id the ring no longer reaches
		// back to, or from another run of this process. Saying so is what
		// lets it fetch afresh instead of quietly showing a fleet that has
		// moved on; the frame carries no id, so the client's own stays put.
		if err := stream.event(string(events.KindResync), "", []byte(`{"reason":"the events since your last id could not all be replayed; fetch the resources again"}`)); err != nil {
			return
		}
	}

	ticker := time.NewTicker(s.streamHeartbeat)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			// The browser navigated away, or the process is shutting down.
			return
		case ev, ok := <-sub.C:
			if !ok {
				// The bus dropped this subscriber for falling behind. Ending
				// the response makes the client reconnect with its last event
				// ID and catch up, which is better than silently going stale.
				return
			}
			data := ev.Data
			if redactPoolEnv && (ev.Kind == events.KindPoolCreated || ev.Kind == events.KindPoolUpdated) {
				data = withoutEnvValues(data)
			}
			if err := stream.event(string(ev.Kind), bus.WireID(ev.ID), data); err != nil {
				return
			}
		case <-ticker.C:
			if s.endIfRevoked(stream, r, auth.ActionEventsRead) {
				return
			}
			if err := stream.comment("heartbeat"); err != nil {
				return
			}
			// The UI's watchdog is driven by heartbeat *events*, not comment
			// frames, so one is sent as well; it carries the server's clock,
			// which is what the top bar's "live" indicator shows.
			if err := stream.event(string(events.KindHeartbeat), "", heartbeatPayload(s)); err != nil {
				return
			}
		}
	}
}

func heartbeatPayload(s *Server) []byte {
	return []byte(`{"at":"` + s.ctrl.Now().Format(time.RFC3339) + `"}`)
}

// lastEventID reads where a reconnecting client left off, and whether that
// place is in this process's sequence at all.
//
// The header is what the browser's own EventSource retry sends. The query
// parameter is what the UI sends when it has given up and opened a fresh
// EventSource, which cannot carry the header; honouring both is what makes a
// reconnect seamless either way. A malformed id means "start from now" rather
// than an error: the alternative is a client that can never reconnect until it
// is reloaded by hand.
func lastEventID(r *http.Request, bus *events.Bus) (id uint64, sameEpoch bool) {
	raw := strings.TrimSpace(r.Header.Get("Last-Event-ID"))
	if raw == "" {
		raw = strings.TrimSpace(r.URL.Query().Get("last_event_id"))
	}
	if raw == "" {
		return 0, true
	}
	id, sameEpoch = bus.ParseWireID(raw)
	if id == 0 && !sameEpoch {
		return 0, true
	}
	return id, sameEpoch
}

func parseKinds(raw string) []events.Kind {
	var out []events.Kind
	for _, k := range strings.Split(raw, ",") {
		if k = strings.TrimSpace(k); k != "" {
			out = append(out, events.Kind(k))
		}
	}
	return out
}

// ---------------------------------------------------------------------------
// GET /api/v1/runners/{id}/logs
// ---------------------------------------------------------------------------

// logChunkKind is the event name a log frame carries. It is not one of the
// fleet event kinds: this is a different stream with a payload of its own.
const logChunkKind = "log"

// handleRunnerLogs tails a runner's output.
//
// The controller never dials an agent, so this asks the controller to have the
// runner's agent open an outbound relay and then fans that out here. The two
// things this handler must get right are ending the stream when the runner's
// output ends, and unsubscribing on every exit -- a leaked subscription keeps a
// relay open on a host for a browser tab that closed hours ago.
func (s *Server) handleRunnerLogs(w http.ResponseWriter, r *http.Request) {
	id := chiURLParam(r, "id")
	opts := backend.LogOptions{
		Follow: queryBool(r, "follow", true),
		Tail:   queryInt(r, "tail", 1000),
	}
	if opts.Tail < 0 {
		opts.Tail = 0
	}

	ctx := r.Context()
	ch, cancel, err := s.ctrl.OpenLogStream(ctx, id, opts)
	if err != nil {
		s.logStreamFailed(w, r, err)
		return
	}
	// Always, on every path out: this is what tells the agent to stop reading
	// the container's output once the last viewer has gone.
	defer cancel()

	stream := startSSE(w, r)
	_ = stream.retry(2 * time.Second)
	if err := stream.comment("attached to " + id); err != nil {
		return
	}

	ticker := time.NewTicker(s.streamHeartbeat)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case chunk, ok := <-ch:
			if !ok {
				// The runner's output finished -- for an ephemeral runner that
				// means the job is over. Say so and end the response, rather
				// than leaving the browser holding a connection that will never
				// carry another byte.
				_ = stream.event("end", "", []byte(`{"reason":"the runner's output ended"}`))
				return
			}
			if err := stream.event(logChunkKind, "", jsonString(string(chunk))); err != nil {
				return
			}
		case <-ticker.C:
			if s.endIfRevoked(stream, r, auth.ActionLogsRead) {
				return
			}
			if err := stream.comment("heartbeat"); err != nil {
				return
			}
		}
	}
}

// handleDownloadRunnerLogs returns a snapshot of a runner's output as a file.
//
// It is the same relay as the live tail, read until it goes quiet rather than
// followed: "download" has to terminate, and a runner that is still producing
// output would otherwise stream forever into a file the browser never finishes.
func (s *Server) handleDownloadRunnerLogs(w http.ResponseWriter, r *http.Request) {
	id := chiURLParam(r, "id")
	run, err := s.ctrl.Store().GetRunner(r.Context(), id)
	if err != nil {
		s.fail(w, r, "reading the runner", err)
		return
	}

	ch, cancel, err := s.ctrl.OpenLogStream(r.Context(), id, backend.LogOptions{Follow: false, Tail: 0})
	if err != nil {
		s.logStreamFailed(w, r, err)
		return
	}
	defer cancel()

	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.Header().Set("Content-Disposition", `attachment; filename="`+logFilename(run.Name)+`"`)
	w.WriteHeader(http.StatusOK)
	rc := http.NewResponseController(w)
	_ = rc.SetWriteDeadline(time.Time{})

	// A backlog arrives in a burst and then stops; this bounds the wait for the
	// first chunk and the gap between chunks, so a wedged agent cannot hold the
	// request open indefinitely.
	const firstChunkWait = 10 * time.Second
	const quietFor = 2 * time.Second

	timer := time.NewTimer(firstChunkWait)
	defer timer.Stop()
	for {
		select {
		case <-r.Context().Done():
			return
		case chunk, ok := <-ch:
			if !ok {
				return
			}
			if _, err := w.Write(chunk); err != nil {
				return
			}
			_ = rc.Flush()
			if !timer.Stop() {
				select {
				case <-timer.C:
				default:
				}
			}
			timer.Reset(quietFor)
		case <-timer.C:
			return
		}
	}
}

// logStreamFailed answers a stream that could not be opened.
//
// A runner that has been removed, or that never reached a host, cannot produce
// output; that is a 409 with the controller's own sentence rather than a 500,
// because it is a fact about the runner and not a fault in the server.
func (s *Server) logStreamFailed(w http.ResponseWriter, r *http.Request, err error) {
	if errors.Is(err, store.ErrNotFound) || errors.Is(err, controller.ErrStreamUnknown) {
		s.fail(w, r, "opening the log stream", err)
		return
	}
	conflict(w, err.Error())
}

// logFilename renders the Content-Disposition name, keeping it to characters
// that survive every filesystem and every quoting rule on the way.
func logFilename(runnerName string) string {
	safe := strings.Map(func(r rune) rune {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9', r == '-', r == '_', r == '.':
			return r
		default:
			return '-'
		}
	}, runnerName)
	if safe == "" {
		safe = "runner"
	}
	return safe + ".log"
}

// jsonString renders a log chunk as a JSON string, which is what keeps a line
// containing a newline or a quote from breaking the frame.
func jsonString(s string) []byte {
	b, err := marshalJSON(s)
	if err != nil {
		return []byte(`""`)
	}
	return b
}

// withoutEnvValues blanks the env values in an already-rendered pool frame,
// keeping every key.
//
// It works on the encoded bytes because that is what the bus carries: a frame
// is rendered once and fanned out to every subscriber, so the alternative --
// rendering per connection -- would make one viewer's presence cost every
// other subscriber a re-render. Only a viewer's connection pays for this, and
// only on a pool frame.
//
// A payload that does not parse is passed through untouched rather than
// dropped: it cannot be a pool view, and swallowing frames would be a worse
// failure than the one this guards against.
func withoutEnvValues(data []byte) []byte {
	var frame map[string]json.RawMessage
	if err := json.Unmarshal(data, &frame); err != nil {
		return data
	}
	raw, ok := frame["env"]
	if !ok {
		return data
	}
	var env map[string]string
	if err := json.Unmarshal(raw, &env); err != nil || len(env) == 0 {
		return data
	}
	for k := range env {
		env[k] = ""
	}
	blanked, err := json.Marshal(env)
	if err != nil {
		return data
	}
	frame["env"] = blanked
	out, err := json.Marshal(frame)
	if err != nil {
		return data
	}
	return out
}

File: web/src/lib/api/sse.ts
/**
 * One shared connection to /api/v1/events for the whole application.
 *
 * There is no refresh button in Zoomies, so this is the only thing standing
 * between the operator and a stale screen. It therefore never throws: a dead
 * connection is a state the top bar renders, not an exception a page has to
 * catch.
 *
 * Reconnection has two paths, deliberately:
 *
 *  * The browser's own EventSource retry, which resends `Last-Event-ID` and so
 *    lets the server replay what it buffered. We leave that alone and just
 *    report `reconnecting` while it happens.
 *  * Our own backoff, for when the browser gives up (a non-2xx response closes
 *    an EventSource for good). Because a fresh EventSource has no last id to
 *    send, we carry it in the query string instead.
 *
 * Either way, regaining focus or coming back online retries immediately rather
 * than waiting out the backoff -- an operator who has just opened the laptop
 * lid should not stare at "reconnecting" for fifteen seconds.
 */
import { api, apiBase } from './client';
import type { EventKind, EventPayloads } from './types';

export type SseStatus = 'connecting' | 'live' | 'reconnecting' | 'offline';

type AnyListener = (data: unknown, id: string | undefined) => void;
type StatusListener = (status: SseStatus) => void;

const BACKOFF_BASE = 500;
const BACKOFF_CAP = 15_000;
/** Server heartbeat is every 20s; three missed ones means something is wrong. */
const WATCHDOG_MS = 70_000;

class EventStream {
  #source: EventSource | null = null;
  #status: SseStatus = 'offline';
  #lastEventId: string | null = null;
  #attempt = 0;
  #retryTimer: ReturnType<typeof setTimeout> | null = null;
  #watchdog: ReturnType<typeof setTimeout> | null = null;
  #heartbeats = false;
  #started = false;
  #bound = false;

  readonly #listeners = new Map<EventKind, Set<AnyListener>>();
  readonly #statusListeners = new Set<StatusListener>();

  get status(): SseStatus {
    return this.#status;
  }

  get lastEventId(): string | null {
    return this.#lastEventId;
  }

  /** Open the stream. Safe to call more than once. */
  start(): void {
    if (this.#started) return;
    this.#started = true;
    this.#bindWindow();
    this.#open();
  }

  /** Close the stream and stop retrying. Used on sign-out. */
  stop(): void {
    this.#started = false;
    this.#clearTimers();
    this.#close();
    this.#setStatus('offline');
  }

  /** Drop the current connection and open a new one now, resetting the backoff. */
  reconnectNow(): void {
    if (!this.#started) return;
    this.#clearTimers();
    this.#close();
    this.#attempt = 0;
    this.#open();
  }

  /**
   * Listen for one or more event kinds. Returns the unsubscribe function; call
   * it from an `$effect` teardown.
   */
  subscribe<K extends EventKind>(
    kinds: K | readonly K[],
    handler: (data: EventPayloads[K], id: string | undefined) => void,
  ): () => void {
    const list = (Array.isArray(kinds) ? kinds : [kinds]) as readonly K[];
    const wrapped: AnyListener = (data, id) => handler(data as EventPayloads[K], id);
    for (const kind of list) {
      let set = this.#listeners.get(kind);
      if (!set) {
        set = new Set();
        this.#listeners.set(kind, set);
      }
      set.add(wrapped);
      if (this.#source) this.#attachKind(this.#source, kind);
    }
    return () => {
      for (const kind of list) this.#listeners.get(kind)?.delete(wrapped);
    };
  }

  /** Listen for connection-state changes. Fires immediately with the current one. */
  onStatus(handler: StatusListener): () => void {
    this.#statusListeners.add(handler);
    handler(this.#status);
    return () => this.#statusListeners.delete(handler);
  }

  /* -- internals --------------------------------------------------------- */

  #bindWindow(): void {
    if (this.#bound || typeof window === 'undefined') return;
    this.#bound = true;
    const wake = () => {
      if (!this.#started) return;
      if (this.#status === 'live') return;
      this.reconnectNow();
    };
    window.addEventListener('online', wake);
    window.addEventListener('focus', wake);
    window.addEventListener('offline', () => {
      if (this.#started) this.#setStatus('offline');
    });
    document.addEventListener('visibilitychange', () => {
      if (document.visibilityState === 'visible') wake();
    });
  }

  #open(): void {
    if (typeof EventSource === 'undefined') return;
    if (typeof navigator !== 'undefined' && navigator.onLine === false) {
      this.#setStatus('offline');
      return;
    }
    this.#setStatus(this.#attempt === 0 ? 'connecting' : 'reconnecting');

    // A fresh EventSource cannot send the Last-Event-ID header, so the id
    // travels in the query string on a manual reconnect. Servers that only
    // read the header ignore it, which costs nothing.
    const query = this.#lastEventId
      ? `?last_event_id=${encodeURIComponent(this.#lastEventId)}`
      : '';
    let source: EventSource;
    try {
      source = new EventSource(`${apiBase}/events${query}`);
    } catch {
      this.#scheduleRetry();
      return;
    }
    this.#source = source;

    source.onopen = () => {
      this.#attempt = 0;
      this.#setStatus('live');
      this.#armWatchdog();
    };
    source.onmessage = (e) => this.#dispatch('heartbeat', e);
    source.onerror = () => {
      // readyState CONNECTING means the browser is retrying by itself and will
      // resend Last-Event-ID; leave it to do that and just report the state.
      if (source.readyState === EventSource.CONNECTING) {
        this.#setStatus('reconnecting');
        return;
      }
      this.#close();
      this.#scheduleRetry();
    };
    for (const kind of this.#listeners.keys()) this.#attachKind(source, kind);
    // Heartbeats are always listened for: they drive the stall watchdog.
    this.#attachKind(source, 'heartbeat');
  }

  #attachKind(source: EventSource, kind: EventKind): void {
    const flag = `zoomies:${kind}`;
    const marked = source as EventSource & { [key: string]: unknown };
    if (marked[flag]) return;
    marked[flag] = true;
    source.addEventListener(kind, (e) => this.#dispatch(kind, e as MessageEvent<string>));
  }

  #dispatch(kind: EventKind, e: MessageEvent<string>): void {
    if (e.lastEventId) this.#lastEventId = e.lastEventId;
    if (kind === 'heartbeat') {
      this.#heartbeats = true;
      this.#armWatchdog();
    } else {
      this.#armWatchdog();
    }
    const listeners = this.#listeners.get(kind);
    if (!listeners || listeners.size === 0) return;
    let data: unknown;
    try {
      data = e.data ? JSON.parse(e.data) : undefined;
    } catch {
      // A malformed frame is the server's problem, not something a page should
      // ever see as an exception. Drop it and keep the stream alive.
      return;
    }
    for (const fn of listeners) {
      try {
        fn(data, e.lastEventId || undefined);
      } catch {
        // A broken subscriber must not take the connection down with it.
      }
    }
  }

  /**
   * Guards against the connection that is open but silent -- a proxy holding a
   * dead socket. Only armed once we have actually seen a heartbeat event, so a
   * server that heartbeats with comment frames does not get reconnected every
   * seventy seconds for no reason.
   */
  #armWatchdog(): void {
    if (this.#watchdog) clearTimeout(this.#watchdog);
    if (!this.#heartbeats) return;
    this.#watchdog = setTimeout(() => {
      if (this.#started) this.reconnectNow();
    }, WATCHDOG_MS);
  }

  /**
   * Find out whether the server is refusing us rather than merely absent.
   *
   * EventSource cannot see why a connection was closed, and a 401 -- the
   * session expiring while the Overview sat on a second monitor -- looks
   * exactly like a controller restart from here: "Reconnecting" for ever
   * over numbers that have quietly stopped moving. One ordinary request per
   * retry answers the question. A 401 runs the client's sign-out hook, which
   * takes the operator to the login page; anything else is left to the retry.
   */
  async #probeSession(): Promise<void> {
    if (typeof navigator !== 'undefined' && navigator.onLine === false) return;
    try {
      await api.get('/auth/session');
    } catch {
      // Either the 401 hook has already acted, or the controller is not
      // answering at all, which the retry is for.
    }
  }

  #scheduleRetry(): void {
    if (!this.#started) return;
    this.#setStatus(navigator?.onLine === false ? 'offline' : 'reconnecting');
    void this.#probeSession();
    const delay = Math.min(BACKOFF_CAP, BACKOFF_BASE * 2 ** this.#attempt);
    // A little jitter so a controller restart does not bring every open dashboard
    // back at exactly the same instant.
    const jittered = delay * (0.75 + Math.random() * 0.5);
    this.#attempt += 1;
    if (this.#retryTimer) clearTimeout(this.#retryTimer);
    this.#retryTimer = setTimeout(() => this.#open(), jittered);
  }

  #close(): void {
    if (this.#source) {
      this.#source.onopen = null;
      this.#source.onerror = null;
      this.#source.onmessage = null;
      this.#source.close();
      this.#source = null;
    }
  }

  #clearTimers(): void {
    if (this.#retryTimer) clearTimeout(this.#retryTimer);
    if (this.#watchdog) clearTimeout(this.#watchdog);
    this.#retryTimer = null;
    this.#watchdog = null;
  }

  #setStatus(next: SseStatus): void {
    if (this.#status === next) return;
    this.#status = next;
    for (const fn of this.#statusListeners) {
      try {
        fn(next);
      } catch {
        // Same reasoning as above.
      }
    }
  }
}

/** The one stream. Started by the shell once the visitor is authenticated. */
export const events = new EventStream();
